common/types/activation.go

```diff
@@ -119,9 +126,24 @@ func (atxh *ActivationTxHeader) TargetEpoch() EpochID {
 	return atxh.PubLayerID.GetEpoch() + 1
 }
 
-// GetWeight of the atx.
+// GetWeight of the ATX. The total weight of the epoch is expected to fit in a uint64 and is
+// sum(atx.NumUnits * atx.TickCount for each ATX in a given epoch).
+// Space Units sizes are chosen such that NumUnits for all ATXs in an epoch is expected to be < 10^6.
+// PoETs should produce ~10k ticks at genesis, but are expected due to technological advances
+// to produce more over time. A uint64 should be large enough to hold the total weight of an epoch,
+// for at least the first few years.
 func (atxh *ActivationTxHeader) GetWeight() uint64 {
-	return uint64(atxh.NumUnits) * (atxh.TickCount)
+	return safeMul(uint64(atxh.NumUnits), (atxh.TickCount))
+}
+
+func safeMul(a, b uint64) uint64 {
+	if a > 0 && b > math.MaxUint64/a {
+		panic("uint64 overflow")
+	}
+	if b > 0 && a > math.MaxUint64/b {
+		panic("uint64 overflow")
+	}
```

Comment:
Why is the second check needed? I assume it's because of rounding down that happens in the division. You could do one check if you did (on line 140) `b >= math.MaxUint64/a + 1`, but all of this is actually really confusing and error prone...

I find [Rob Pike's version](https://groups.google.com/g/golang-nuts/c/h5oSN5t3Au4/m/KaNQREhZh0QJ) with a single division more intuitively obvious - I need to think less to see that it's safe.

Adapted here it would be:
```go
func safeMul(a, b uint64) uint64 {
	c := a * b
	if a > 1 && b > 1 && c/b != a {
		panic("uint64 overflow")
	}
	return c
}
```

I'd be curious to see a benchmark comparison of the 3 options (and also to make sure that my version passes tests - maybe I'm missing something).

Reply:
A very quick benchmark shows that this is not only faster than the double check version, but also hardly slower than multiplying without overflow check:

```
goos: linux
goarch: arm64
pkg: github.com/spacemeshos/go-spacemesh/common/types
BenchmarkSafeMul-6    	940181595	         1.256 ns/op	       0 B/op	       0 allocs/op
BenchmarkSafeMul2-6   	1000000000	         0.6292 ns/op	       0 B/op	       0 allocs/op
BenchmarkMul-6            	1000000000	         0.4933 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/spacemeshos/go-spacemesh/common/types	2.459s
```

I'll update the implementation to this version.